From: Alexander Mikhaylenko <alexm@gnome.org>
Date: Wed, 24 Feb 2021 19:37:24 +0500
Subject: Disable fullscreen

We'll need to disable fullscreen on phosh because it doesn't work there.
This isn't a matter of adaptiveness, because fullscreen technically can
work on mobile, but just doesn't in phosh. So let's add a check.
---
 embed/ephy-embed-prefs.c               | 13 +++++++++++++
 src/ephy-desktop-utils.c               | 11 +++++++++++
 src/ephy-desktop-utils.h               |  1 +
 src/ephy-header-bar.c                  |  5 +++++
 src/ephy-lockdown.c                    |  3 ++-
 src/ephy-window.c                      |  2 +-
 src/resources/gtk/page-menu-popover.ui |  2 +-
 src/window-commands.c                  |  3 ++-
 8 files changed, 36 insertions(+), 4 deletions(-)

diff --git a/embed/ephy-embed-prefs.c b/embed/ephy-embed-prefs.c
index 6302186..d55f286 100644
--- a/embed/ephy-embed-prefs.c
+++ b/embed/ephy-embed-prefs.c
@@ -558,12 +558,25 @@ static const PrefData webkit_pref_entries[] = {
     webkit_pref_callback_hardware_acceleration_policy },
 };
 
+/* Unfortunately we can't reuse ephy-desktop-utils.h here */
+static gboolean
+is_session_phosh (void)
+{
+  const gchar *xdg_session_desktop = g_getenv ("XDG_SESSION_DESKTOP");
+
+  if (!xdg_session_desktop)
+    return FALSE;
+
+  return strstr (xdg_session_desktop, "phosh") != NULL;
+}
+
 static gpointer
 ephy_embed_prefs_init (gpointer user_data)
 {
   guint i;
 
   webkit_settings = webkit_settings_new_with_settings ("enable-back-forward-navigation-gestures", TRUE,
+                                                       "enable-fullscreen", !is_session_phosh (),
                                                        "enable-developer-extras", TRUE,
                                                        "enable-dns-prefetching", TRUE,
                                                        "enable-plugins", FALSE,
diff --git a/src/ephy-desktop-utils.c b/src/ephy-desktop-utils.c
index 5f52bb7..d41c4c4 100644
--- a/src/ephy-desktop-utils.c
+++ b/src/ephy-desktop-utils.c
@@ -46,6 +46,17 @@ is_desktop_gnome (void)
   return strstr (xdg_current_desktop, "GNOME") != NULL;
 }
 
+gboolean
+is_session_phosh (void)
+{
+  const gchar *xdg_session_desktop = g_getenv ("XDG_SESSION_DESKTOP");
+
+  if (!xdg_session_desktop)
+    return FALSE;
+
+  return strstr (xdg_session_desktop, "phosh") != NULL;
+}
+
 GtkIconSize
 get_icon_size (void)
 {
diff --git a/src/ephy-desktop-utils.h b/src/ephy-desktop-utils.h
index c162f3f..64f739b 100644
--- a/src/ephy-desktop-utils.h
+++ b/src/ephy-desktop-utils.h
@@ -32,6 +32,7 @@ typedef enum {
 
 gboolean is_desktop_pantheon (void);
 gboolean is_desktop_gnome (void);
+gboolean is_session_phosh (void);
 
 GtkIconSize get_icon_size (void);
 
diff --git a/src/ephy-header-bar.c b/src/ephy-header-bar.c
index 9af2d39..f9b17e2 100644
--- a/src/ephy-header-bar.c
+++ b/src/ephy-header-bar.c
@@ -331,6 +331,11 @@ ephy_header_bar_constructed (GObject *object)
   }
   g_settings_bind (EPHY_SETTINGS_WEB, EPHY_PREFS_WEB_ENABLE_WEBEXTENSIONS, gtk_builder_get_object (builder, "extensions-button"), "visible", G_SETTINGS_BIND_DEFAULT);
 
+  if (is_session_phosh ()) {
+    GtkWidget *fullscreen_button = GTK_WIDGET (gtk_builder_get_object (builder, "fullscreen_button"));
+    gtk_widget_hide (fullscreen_button);
+  }
+
   gtk_menu_button_set_popover (GTK_MENU_BUTTON (button), header_bar->page_menu_popover);
   g_object_unref (builder);
 
diff --git a/src/ephy-lockdown.c b/src/ephy-lockdown.c
index 19d35af..4cc0700 100644
--- a/src/ephy-lockdown.c
+++ b/src/ephy-lockdown.c
@@ -25,6 +25,7 @@
 
 #include "ephy-action-helper.h"
 #include "ephy-debug.h"
+#include "ephy-desktop-utils.h"
 #include "ephy-embed-container.h"
 #include "ephy-embed-shell.h"
 #include "ephy-embed-utils.h"
@@ -70,7 +71,7 @@ fullscreen_cb (GSettings  *settings,
                const char *key,
                EphyWindow *window)
 {
-  if (g_settings_get_boolean (settings, key))
+  if (g_settings_get_boolean (settings, key) && !is_session_phosh ())
     gtk_window_fullscreen (GTK_WINDOW (window));
   else
     gtk_window_unfullscreen (GTK_WINDOW (window));
diff --git a/src/ephy-window.c b/src/ephy-window.c
index 857ac81..8bab297 100644
--- a/src/ephy-window.c
+++ b/src/ephy-window.c
@@ -3141,7 +3141,7 @@ ephy_window_state_event (GtkWidget           *widget,
   EphyWindow *window = EPHY_WINDOW (widget);
   gboolean result = GDK_EVENT_PROPAGATE;
 
-  if (event->changed_mask & GDK_WINDOW_STATE_FULLSCREEN) {
+  if (event->changed_mask & GDK_WINDOW_STATE_FULLSCREEN && !is_session_phosh ()) {
     GActionGroup *action_group;
     GAction *action;
     gboolean fullscreen;
diff --git a/src/resources/gtk/page-menu-popover.ui b/src/resources/gtk/page-menu-popover.ui
index 25df01a..ae86a5d 100644
--- a/src/resources/gtk/page-menu-popover.ui
+++ b/src/resources/gtk/page-menu-popover.ui
@@ -115,7 +115,7 @@
               </packing>
             </child>
             <child>
-              <object class="GtkModelButton">
+              <object class="GtkModelButton" id="fullscreen_button">
                 <property name="tooltip_text" translatable="yes">Fullscreen</property>
                 <property name="action-name">win.fullscreen</property>
                 <property name="iconic">True</property>
diff --git a/src/window-commands.c b/src/window-commands.c
index 6f46496..9b153d1 100644
--- a/src/window-commands.c
+++ b/src/window-commands.c
@@ -30,6 +30,7 @@
 #include "ephy-bookmarks-import.h"
 #include "ephy-bookmarks-manager.h"
 #include "ephy-debug.h"
+#include "ephy-desktop-utils.h"
 #include "ephy-embed-container.h"
 #include "ephy-embed-prefs.h"
 #include "ephy-embed-shell.h"
@@ -2641,7 +2642,7 @@ window_cmd_change_fullscreen_state (GSimpleAction *action,
   EphyWindow *window = EPHY_WINDOW (user_data);
   gboolean active;
 
-  active = g_variant_get_boolean (state);
+  active = g_variant_get_boolean (state) && !is_session_phosh ();
 
   /* This is performed only here because we don't want it occuring when a window
    * enters fullscreen mode for some other reason other than action activation.
